File: src/main/java/com/merge/api/resources/crm/types/AssociationType.java
package com.merge.api.resources.crm.types;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.JsonSetter;
import com.fasterxml.jackson.annotation.Nulls;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import com.merge.api.core.ObjectMappers;
import java.time.OffsetDateTime;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Optional;

@JsonInclude(JsonInclude.Include.NON_EMPTY)
@JsonDeserialize(builder = AssociationType.Builder.class)
public final class AssociationType {
    private final Optional<Map<String, JsonNode>> sourceObjectClass;

    private final Optional<List<AssociationSubType>> targetObjectClasses;

    private final Optional<String> remoteKeyName;

    private final Optional<String> displayName;

    private final Optional<AssociationTypeCardinality> cardinality;

    private final Optional<Boolean> isRequired;

    private final Optional<String> id;

    private final Optional<String> remoteId;

    private final Optional<OffsetDateTime> createdAt;

    private final Optional<OffsetDateTime> modifiedAt;

    private AssociationType(
            Optional<Map<String, JsonNode>> sourceObjectClass,
            Optional<List<AssociationSubType>> targetObjectClasses,
            Optional<String> remoteKeyName,
            Optional<String> displayName,
            Optional<AssociationTypeCardinality> cardinality,
            Optional<Boolean> isRequired,
            Optional<String> id,
            Optional<String> remoteId,
            Optional<OffsetDateTime> createdAt,
            Optional<OffsetDateTime> modifiedAt) {
        this.sourceObjectClass = sourceObjectClass;
        this.targetObjectClasses = targetObjectClasses;
        this.remoteKeyName = remoteKeyName;
        this.displayName = displayName;
        this.cardinality = cardinality;
        this.isRequired = isRequired;
        this.id = id;
        this.remoteId = remoteId;
        this.createdAt = createdAt;
        this.modifiedAt = modifiedAt;
    }

    @JsonProperty("source_object_class")
    public Optional<Map<String, JsonNode>> getSourceObjectClass() {
        return sourceObjectClass;
    }

    @JsonProperty("target_object_classes")
    public Optional<List<AssociationSubType>> getTargetObjectClasses() {
        return targetObjectClasses;
    }

    @JsonProperty("remote_key_name")
    public Optional<String> getRemoteKeyName() {
        return remoteKeyName;
    }

    @JsonProperty("display_name")
    public Optional<String> getDisplayName() {
        return displayName;
    }

    @JsonProperty("cardinality")
    public Optional<AssociationTypeCardinality> getCardinality() {
        return cardinality;
    }

    @JsonProperty("is_required")
    public Optional<Boolean> getIsRequired() {
        return isRequired;
    }

    @JsonProperty("id")
    public Optional<String> getId() {
        return id;
    }

    /**
     * @return The third-party API ID of the matching object.
     */
    @JsonProperty("remote_id")
    public Optional<String> getRemoteId() {
        return remoteId;
    }

    @JsonProperty("created_at")
    public Optional<OffsetDateTime> getCreatedAt() {
        return createdAt;
    }

    /**
     * @return This is the datetime that this object was last updated by Merge
     */
    @JsonProperty("modified_at")
    public Optional<OffsetDateTime> getModifiedAt() {
        return modifiedAt;
    }

    @Override
    public boolean equals(Object other) {
        if (this == other) return true;
        return other instanceof AssociationType && equalTo((AssociationType) other);
    }

    private boolean equalTo(AssociationType other) {
        return sourceObjectClass.equals(other.sourceObjectClass)
                && targetObjectClasses.equals(other.targetObjectClasses)
                && remoteKeyName.equals(other.remoteKeyName)
                && displayName.equals(other.displayName)
                && cardinality.equals(other.cardinality)
                && isRequired.equals(other.isRequired)
                && id.equals(other.id)
                && remoteId.equals(other.remoteId)
                && createdAt.equals(other.createdAt)
                && modifiedAt.equals(other.modifiedAt);
    }

    @Override
    public int hashCode() {
        return Objects.hash(
                this.sourceObjectClass,
                this.targetObjectClasses,
                this.remoteKeyName,
                this.displayName,
                this.cardinality,
                this.isRequired,
                this.id,
                this.remoteId,
                this.createdAt,
                this.modifiedAt);
    }

    @Override
    public String toString() {
        return ObjectMappers.stringify(this);
    }

    public static Builder builder() {
        return new Builder();
    }

    @JsonIgnoreProperties(ignoreUnknown = true)
    public static final class Builder {
        private Optional<Map<String, JsonNode>> sourceObjectClass = Optional.empty();

        private Optional<List<AssociationSubType>> targetObjectClasses = Optional.empty();

        private Optional<String> remoteKeyName = Optional.empty();

        private Optional<String> displayName = Optional.empty();

        private Optional<AssociationTypeCardinality> cardinality = Optional.empty();

        private Optional<Boolean> isRequired = Optional.empty();

        private Optional<String> id = Optional.empty();

        private Optional<String> remoteId = Optional.empty();

        private Optional<OffsetDateTime> createdAt = Optional.empty();

        private Optional<OffsetDateTime> modifiedAt = Optional.empty();

        private Builder() {}

        public Builder from(AssociationType other) {
            sourceObjectClass(other.getSourceObjectClass());
            targetObjectClasses(other.getTargetObjectClasses());
            remoteKeyName(other.getRemoteKeyName());
            displayName(other.getDisplayName());
            cardinality(other.getCardinality());
            isRequired(other.getIsRequired());
            id(other.getId());
            remoteId(other.getRemoteId());
            createdAt(other.getCreatedAt());
            modifiedAt(other.getModifiedAt());
            return this;
        }

        @JsonSetter(value = "source_object_class", nulls = Nulls.SKIP)
        public Builder sourceObjectClass(Optional<Map<String, JsonNode>> sourceObjectClass) {
            this.sourceObjectClass = sourceObjectClass;
            return this;
        }

        public Builder sourceObjectClass(Map<String, JsonNode> sourceObjectClass) {
            this.sourceObjectClass = Optional.of(sourceObjectClass);
            return this;
        }

        @JsonSetter(value = "target_object_classes", nulls = Nulls.SKIP)
        public Builder targetObjectClasses(Optional<List<AssociationSubType>> targetObjectClasses) {
            this.targetObjectClasses = targetObjectClasses;
            return this;
        }

        public Builder targetObjectClasses(List<AssociationSubType> targetObjectClasses) {
            this.targetObjectClasses = Optional.of(targetObjectClasses);
            return this;
        }

        @JsonSetter(value = "remote_key_name", nulls = Nulls.SKIP)
        public Builder remoteKeyName(Optional<String> remoteKeyName) {
            this.remoteKeyName = remoteKeyName;
            return this;
        }

        public Builder remoteKeyName(String remoteKeyName) {
            this.remoteKeyName = Optional.of(remoteKeyName);
            return this;
        }

        @JsonSetter(value = "display_name", nulls = Nulls.SKIP)
        public Builder displayName(Optional<String> displayName) {
            this.displayName = displayName;
            return this;
        }

        public Builder displayName(String displayName) {
            this.displayName = Optional.of(displayName);
            return this;
        }

        @JsonSetter(value = "cardinality", nulls = Nulls.SKIP)
        public Builder cardinality(Optional<AssociationTypeCardinality> cardinality) {
            this.cardinality = cardinality;
            return this;
        }

        public Builder cardinality(AssociationTypeCardinality cardinality) {
            this.cardinality = Optional.of(cardinality);
            return this;
        }

        @JsonSetter(value = "is_required", nulls = Nulls.SKIP)
        public Builder isRequired(Optional<Boolean> isRequired) {
            this.isRequired = isRequired;
            return this;
        }

        public Builder isRequired(Boolean isRequired) {
            this.isRequired = Optional.of(isRequired);
            return this;
        }

        @JsonSetter(value = "id", nulls = Nulls.SKIP)
        public Builder id(Optional<String> id) {
            this.id = id;
            return this;
        }

        public Builder id(String id) {
            this.id = Optional.of(id);
            return this;
        }

        @JsonSetter(value = "remote_id", nulls = Nulls.SKIP)
        public Builder remoteId(Optional<String> remoteId) {
            this.remoteId = remoteId;
            return this;
        }

        public Builder remoteId(String remoteId) {
            this.remoteId = Optional.of(remoteId);
            return this;
        }

        @JsonSetter(value = "created_at", nulls = Nulls.SKIP)
        public Builder createdAt(Optional<OffsetDateTime> createdAt) {
            this.createdAt = createdAt;
            return this;
        }

        public Builder createdAt(OffsetDateTime createdAt) {
            this.createdAt = Optional.of(createdAt);
            return this;
        }

        @JsonSetter(value = "modified_at", nulls = Nulls.SKIP)
        public Builder modifiedAt(Optional<OffsetDateTime> modifiedAt) {
            this.modifiedAt = modifiedAt;
            return this;
        }

        public Builder modifiedAt(OffsetDateTime modifiedAt) {
            this.modifiedAt = Optional.of(modifiedAt);
            return this;
        }

        public AssociationType build() {
            return new AssociationType(
                    sourceObjectClass,
                    targetObjectClasses,
                    remoteKeyName,
                    displayName,
                    cardinality,
                    isRequired,
                    id,
                    remoteId,
                    createdAt,
                    modifiedAt);
        }
    }
}

File: merge-java-core/src/main/kotlin/dev/merge/api/models/crm/AssociationType.kt
package dev.merge.api.models.crm

import com.fasterxml.jackson.annotation.JsonAnyGetter
import com.fasterxml.jackson.annotation.JsonAnySetter
import com.fasterxml.jackson.annotation.JsonCreator
import com.fasterxml.jackson.annotation.JsonProperty
import com.fasterxml.jackson.databind.annotation.JsonDeserialize
import dev.merge.api.core.ExcludeMissing
import dev.merge.api.core.JsonField
import dev.merge.api.core.JsonMissing
import dev.merge.api.core.JsonValue
import dev.merge.api.core.NoAutoDetect
import dev.merge.api.core.toUnmodifiable
import dev.merge.api.errors.MergeInvalidDataException
import dev.merge.api.models.*
import java.time.OffsetDateTime
import java.util.Objects
import java.util.Optional

/**
 * # The AssociationType Object
 *
 * ### Description
 *
 * The `Association Type` object represents the relationship between two objects.
 *
 * ### Usage Example
 *
 * TODO
 */
@JsonDeserialize(builder = AssociationType.Builder::class)
@NoAutoDetect
class AssociationType
private constructor(
    private val sourceObjectClass: JsonValue,
    private val targetObjectClasses: JsonField<List<AssociationSubType>>,
    private val remoteKeyName: JsonField<String>,
    private val displayName: JsonField<String>,
    private val cardinality: JsonField<CardinalityEnum>,
    private val isRequired: JsonField<Boolean>,
    private val id: JsonField<String>,
    private val remoteId: JsonField<String>,
    private val modifiedAt: JsonField<OffsetDateTime>,
    private val additionalProperties: Map<String, JsonValue>,
) {

    private var validated: Boolean = false

    private var hashCode: Int = 0

    fun targetObjectClasses(): Optional<List<AssociationSubType>> =
        Optional.ofNullable(targetObjectClasses.getNullable("target_object_classes"))

    fun remoteKeyName(): Optional<String> =
        Optional.ofNullable(remoteKeyName.getNullable("remote_key_name"))

    fun displayName(): Optional<String> =
        Optional.ofNullable(displayName.getNullable("display_name"))

    /**
     * - `ONE_TO_ONE` - ONE_TO_ONE
     * - `MANY_TO_ONE` - MANY_TO_ONE
     * - `MANY_TO_MANY` - MANY_TO_MANY
     * - `ONE_TO_MANY` - ONE_TO_MANY
     */
    fun cardinality(): Optional<CardinalityEnum> =
        Optional.ofNullable(cardinality.getNullable("cardinality"))

    fun isRequired(): Optional<Boolean> = Optional.ofNullable(isRequired.getNullable("is_required"))

    fun id(): Optional<String> = Optional.ofNullable(id.getNullable("id"))

    /** The third-party API ID of the matching object. */
    fun remoteId(): Optional<String> = Optional.ofNullable(remoteId.getNullable("remote_id"))

    /** This is the datetime that this object was last updated by Merge */
    fun modifiedAt(): Optional<OffsetDateTime> =
        Optional.ofNullable(modifiedAt.getNullable("modified_at"))

    @JsonProperty("source_object_class")
    @ExcludeMissing
    fun _sourceObjectClass() = sourceObjectClass

    @JsonProperty("target_object_classes")
    @ExcludeMissing
    fun _targetObjectClasses() = targetObjectClasses

    @JsonProperty("remote_key_name") @ExcludeMissing fun _remoteKeyName() = remoteKeyName

    @JsonProperty("display_name") @ExcludeMissing fun _displayName() = displayName

    /**
     * - `ONE_TO_ONE` - ONE_TO_ONE
     * - `MANY_TO_ONE` - MANY_TO_ONE
     * - `MANY_TO_MANY` - MANY_TO_MANY
     * - `ONE_TO_MANY` - ONE_TO_MANY
     */
    @JsonProperty("cardinality") @ExcludeMissing fun _cardinality() = cardinality

    @JsonProperty("is_required") @ExcludeMissing fun _isRequired() = isRequired

    @JsonProperty("id") @ExcludeMissing fun _id() = id

    /** The third-party API ID of the matching object. */
    @JsonProperty("remote_id") @ExcludeMissing fun _remoteId() = remoteId

    /** This is the datetime that this object was last updated by Merge */
    @JsonProperty("modified_at") @ExcludeMissing fun _modifiedAt() = modifiedAt

    @JsonAnyGetter
    @ExcludeMissing
    fun _additionalProperties(): Map<String, JsonValue> = additionalProperties

    fun validate() = apply {
        if (!validated) {
            targetObjectClasses().map { it.forEach { it.validate() } }
            remoteKeyName()
            displayName()
            cardinality()
            isRequired()
            id()
            remoteId()
            modifiedAt()
            validated = true
        }
    }

    fun toBuilder() = Builder().from(this)

    override fun equals(other: Any?): Boolean {
        if (this === other) {
            return true
        }

        return other is AssociationType &&
            this.sourceObjectClass == other.sourceObjectClass &&
            this.targetObjectClasses == other.targetObjectClasses &&
            this.remoteKeyName == other.remoteKeyName &&
            this.displayName == other.displayName &&
            this.cardinality == other.cardinality &&
            this.isRequired == other.isRequired &&
            this.id == other.id &&
            this.remoteId == other.remoteId &&
            this.modifiedAt == other.modifiedAt &&
            this.additionalProperties == other.additionalProperties
    }

    override fun hashCode(): Int {
        if (hashCode == 0) {
            hashCode =
                Objects.hash(
                    sourceObjectClass,
                    targetObjectClasses,
                    remoteKeyName,
                    displayName,
                    cardinality,
                    isRequired,
                    id,
                    remoteId,
                    modifiedAt,
                    additionalProperties,
                )
        }
        return hashCode
    }

    override fun toString() =
        "AssociationType{sourceObjectClass=$sourceObjectClass, targetObjectClasses=$targetObjectClasses, remoteKeyName=$remoteKeyName, displayName=$displayName, cardinality=$cardinality, isRequired=$isRequired, id=$id, remoteId=$remoteId, modifiedAt=$modifiedAt, additionalProperties=$additionalProperties}"

    companion object {

        @JvmStatic fun builder() = Builder()
    }

    class Builder {

        private var sourceObjectClass: JsonValue = JsonMissing.of()
        private var targetObjectClasses: JsonField<List<AssociationSubType>> = JsonMissing.of()
        private var remoteKeyName: JsonField<String> = JsonMissing.of()
        private var displayName: JsonField<String> = JsonMissing.of()
        private var cardinality: JsonField<CardinalityEnum> = JsonMissing.of()
        private var isRequired: JsonField<Boolean> = JsonMissing.of()
        private var id: JsonField<String> = JsonMissing.of()
        private var remoteId: JsonField<String> = JsonMissing.of()
        private var modifiedAt: JsonField<OffsetDateTime> = JsonMissing.of()
        private var additionalProperties: MutableMap<String, JsonValue> = mutableMapOf()

        @JvmSynthetic
        internal fun from(associationType: AssociationType) = apply {
            this.sourceObjectClass = associationType.sourceObjectClass
            this.targetObjectClasses = associationType.targetObjectClasses
            this.remoteKeyName = associationType.remoteKeyName
            this.displayName = associationType.displayName
            this.cardinality = associationType.cardinality
            this.isRequired = associationType.isRequired
            this.id = associationType.id
            this.remoteId = associationType.remoteId
            this.modifiedAt = associationType.modifiedAt
            additionalProperties(associationType.additionalProperties)
        }

        @JsonProperty("source_object_class")
        @ExcludeMissing
        fun sourceObjectClass(sourceObjectClass: JsonValue) = apply {
            this.sourceObjectClass = sourceObjectClass
        }

        fun targetObjectClasses(targetObjectClasses: List<AssociationSubType>) =
            targetObjectClasses(JsonField.of(targetObjectClasses))

        @JsonProperty("target_object_classes")
        @ExcludeMissing
        fun targetObjectClasses(targetObjectClasses: JsonField<List<AssociationSubType>>) = apply {
            this.targetObjectClasses = targetObjectClasses
        }

        fun remoteKeyName(remoteKeyName: String) = remoteKeyName(JsonField.of(remoteKeyName))

        @JsonProperty("remote_key_name")
        @ExcludeMissing
        fun remoteKeyName(remoteKeyName: JsonField<String>) = apply {
            this.remoteKeyName = remoteKeyName
        }

        fun displayName(displayName: String) = displayName(JsonField.of(displayName))

        @JsonProperty("display_name")
        @ExcludeMissing
        fun displayName(displayName: JsonField<String>) = apply { this.displayName = displayName }

        /**
         * - `ONE_TO_ONE` - ONE_TO_ONE
         * - `MANY_TO_ONE` - MANY_TO_ONE
         * - `MANY_TO_MANY` - MANY_TO_MANY
         * - `ONE_TO_MANY` - ONE_TO_MANY
         */
        fun cardinality(cardinality: CardinalityEnum) = cardinality(JsonField.of(cardinality))

        /**
         * - `ONE_TO_ONE` - ONE_TO_ONE
         * - `MANY_TO_ONE` - MANY_TO_ONE
         * - `MANY_TO_MANY` - MANY_TO_MANY
         * - `ONE_TO_MANY` - ONE_TO_MANY
         */
        @JsonProperty("cardinality")
        @ExcludeMissing
        fun cardinality(cardinality: JsonField<CardinalityEnum>) = apply {
            this.cardinality = cardinality
        }

        fun isRequired(isRequired: Boolean) = isRequired(JsonField.of(isRequired))

        @JsonProperty("is_required")
        @ExcludeMissing
        fun isRequired(isRequired: JsonField<Boolean>) = apply { this.isRequired = isRequired }

        fun id(id: String) = id(JsonField.of(id))

        @JsonProperty("id") @ExcludeMissing fun id(id: JsonField<String>) = apply { this.id = id }

        /** The third-party API ID of the matching object. */
        fun remoteId(remoteId: String) = remoteId(JsonField.of(remoteId))

        /** The third-party API ID of the matching object. */
        @JsonProperty("remote_id")
        @ExcludeMissing
        fun remoteId(remoteId: JsonField<String>) = apply { this.remoteId = remoteId }

        /** This is the datetime that this object was last updated by Merge */
        fun modifiedAt(modifiedAt: OffsetDateTime) = modifiedAt(JsonField.of(modifiedAt))

        /** This is the datetime that this object was last updated by Merge */
        @JsonProperty("modified_at")
        @ExcludeMissing
        fun modifiedAt(modifiedAt: JsonField<OffsetDateTime>) = apply {
            this.modifiedAt = modifiedAt
        }

        fun additionalProperties(additionalProperties: Map<String, JsonValue>) = apply {
            this.additionalProperties.clear()
            this.additionalProperties.putAll(additionalProperties)
        }

        @JsonAnySetter
        fun putAdditionalProperty(key: String, value: JsonValue) = apply {
            this.additionalProperties.put(key, value)
        }

        fun putAllAdditionalProperties(additionalProperties: Map<String, JsonValue>) = apply {
            this.additionalProperties.putAll(additionalProperties)
        }

        fun build(): AssociationType =
            AssociationType(
                sourceObjectClass,
                targetObjectClasses.map { it.toUnmodifiable() },
                remoteKeyName,
                displayName,
                cardinality,
                isRequired,
                id,
                remoteId,
                modifiedAt,
                additionalProperties.toUnmodifiable(),
            )
    }

    @JsonDeserialize(builder = AssociationSubType.Builder::class)
    @NoAutoDetect
    class AssociationSubType
    private constructor(
        private val id: JsonField<String>,
        private val originType: JsonField<String>,
        private val modifiedAt: JsonField<OffsetDateTime>,
        private val additionalProperties: Map<String, JsonValue>,
    ) {

        private var validated: Boolean = false

        private var hashCode: Int = 0

        fun id(): Optional<String> = Optional.ofNullable(id.getNullable("id"))

        fun originType(): Optional<String> =
            Optional.ofNullable(originType.getNullable("origin_type"))

        /** This is the datetime that this object was last updated by Merge */
        fun modifiedAt(): Optional<OffsetDateTime> =
            Optional.ofNullable(modifiedAt.getNullable("modified_at"))

        @JsonProperty("id") @ExcludeMissing fun _id() = id

        @JsonProperty("origin_type") @ExcludeMissing fun _originType() = originType

        /** This is the datetime that this object was last updated by Merge */
        @JsonProperty("modified_at") @ExcludeMissing fun _modifiedAt() = modifiedAt

        @JsonAnyGetter
        @ExcludeMissing
        fun _additionalProperties(): Map<String, JsonValue> = additionalProperties

        fun validate() = apply {
            if (!validated) {
                id()
                originType()
                modifiedAt()
                validated = true
            }
        }

        fun toBuilder() = Builder().from(this)

        override fun equals(other: Any?): Boolean {
            if (this === other) {
                return true
            }

            return other is AssociationSubType &&
                this.id == other.id &&
                this.originType == other.originType &&
                this.modifiedAt == other.modifiedAt &&
                this.additionalProperties == other.additionalProperties
        }

        override fun hashCode(): Int {
            if (hashCode == 0) {
                hashCode =
                    Objects.hash(
                        id,
                        originType,
                        modifiedAt,
                        additionalProperties,
                    )
            }
            return hashCode
        }

        override fun toString() =
            "AssociationSubType{id=$id, originType=$originType, modifiedAt=$modifiedAt, additionalProperties=$additionalProperties}"

        companion object {

            @JvmStatic fun builder() = Builder()
        }

        class Builder {

            private var id: JsonField<String> = JsonMissing.of()
            private var originType: JsonField<String> = JsonMissing.of()
            private var modifiedAt: JsonField<OffsetDateTime> = JsonMissing.of()
            private var additionalProperties: MutableMap<String, JsonValue> = mutableMapOf()

            @JvmSynthetic
            internal fun from(associationSubType: AssociationSubType) = apply {
                this.id = associationSubType.id
                this.originType = associationSubType.originType
                this.modifiedAt = associationSubType.modifiedAt
                additionalProperties(associationSubType.additionalProperties)
            }

            fun id(id: String) = id(JsonField.of(id))

            @JsonProperty("id")
            @ExcludeMissing
            fun id(id: JsonField<String>) = apply { this.id = id }

            fun originType(originType: String) = originType(JsonField.of(originType))

            @JsonProperty("origin_type")
            @ExcludeMissing
            fun originType(originType: JsonField<String>) = apply { this.originType = originType }

            /** This is the datetime that this object was last updated by Merge */
            fun modifiedAt(modifiedAt: OffsetDateTime) = modifiedAt(JsonField.of(modifiedAt))

            /** This is the datetime that this object was last updated by Merge */
            @JsonProperty("modified_at")
            @ExcludeMissing
            fun modifiedAt(modifiedAt: JsonField<OffsetDateTime>) = apply {
                this.modifiedAt = modifiedAt
            }

            fun additionalProperties(additionalProperties: Map<String, JsonValue>) = apply {
                this.additionalProperties.clear()
                this.additionalProperties.putAll(additionalProperties)
            }

            @JsonAnySetter
            fun putAdditionalProperty(key: String, value: JsonValue) = apply {
                this.additionalProperties.put(key, value)
            }

            fun putAllAdditionalProperties(additionalProperties: Map<String, JsonValue>) = apply {
                this.additionalProperties.putAll(additionalProperties)
            }

            fun build(): AssociationSubType =
                AssociationSubType(
                    id,
                    originType,
                    modifiedAt,
                    additionalProperties.toUnmodifiable(),
                )
        }
    }

    class CardinalityEnum
    @JsonCreator
    private constructor(
        private val value: JsonField<String>,
    ) {

        @com.fasterxml.jackson.annotation.JsonValue fun _value(): JsonField<String> = value

        override fun equals(other: Any?): Boolean {
            if (this === other) {
                return true
            }

            return other is CardinalityEnum && this.value == other.value
        }

        override fun hashCode() = value.hashCode()

        override fun toString() = value.toString()

        companion object {

            @JvmField val ONE_TO_ONE = CardinalityEnum(JsonField.of("ONE_TO_ONE"))

            @JvmField val MANY_TO_ONE = CardinalityEnum(JsonField.of("MANY_TO_ONE"))

            @JvmField val MANY_TO_MANY = CardinalityEnum(JsonField.of("MANY_TO_MANY"))

            @JvmField val ONE_TO_MANY = CardinalityEnum(JsonField.of("ONE_TO_MANY"))

            @JvmStatic fun of(value: String) = CardinalityEnum(JsonField.of(value))
        }

        enum class Known {
            ONE_TO_ONE,
            MANY_TO_ONE,
            MANY_TO_MANY,
            ONE_TO_MANY,
        }

        enum class Value {
            ONE_TO_ONE,
            MANY_TO_ONE,
            MANY_TO_MANY,
            ONE_TO_MANY,
            _UNKNOWN,
        }

        fun value(): Value =
            when (this) {
                ONE_TO_ONE -> Value.ONE_TO_ONE
                MANY_TO_ONE -> Value.MANY_TO_ONE
                MANY_TO_MANY -> Value.MANY_TO_MANY
                ONE_TO_MANY -> Value.ONE_TO_MANY
                else -> Value._UNKNOWN
            }

        fun known(): Known =
            when (this) {
                ONE_TO_ONE -> Known.ONE_TO_ONE
                MANY_TO_ONE -> Known.MANY_TO_ONE
                MANY_TO_MANY -> Known.MANY_TO_MANY
                ONE_TO_MANY -> Known.ONE_TO_MANY
                else -> throw MergeInvalidDataException("Unknown CardinalityEnum: $value")
            }

        fun asString(): String = _value().asStringOrThrow()
    }
}
